src/main/java/org/testng/internal/Invoker.java

```diff
@@ -88,12 +88,14 @@
   private static final Predicate<ITestNGMethod, IClass> SAME_CLASS = new SameClassNamePredicate();
 
   private void setClassInvocationFailure(Class<?> clazz, Object instance) {
-    Set<Object> instances = m_classInvocationResults.get( clazz );
-    if (instances == null) {
-      instances = Sets.newHashSet();
-      m_classInvocationResults.put(clazz, instances);
+    synchronized(this){
```

Comment:
As I know, `synchronized(this)` is not a good practice, and in your case, you just want a synchronization on `m_classInvocationResults`.